integration-tests/combined-extension/lib.rs

```diff
@@ -0,0 +1,209 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+use ink::{
+    env::{
+        DefaultEnvironment,
+        Environment,
+    },
+    ChainExtensionInstance,
+};
+use psp22_extension::Psp22Extension;
+use rand_extension::{
+    FetchRandom,
+    RandomReadErr,
+};
+
+/// This extension combines the [`FetchRandom`] and [`Psp22Extension`] extensions.
+/// It implements the [`ChainExtensionInstance`] trait by simply calling
+/// the corresponding sub-extension implementation.
+/// It is possible to combine any number of extensions in this way.
+#[ink::scale_derive(TypeInfo)]
+pub struct CombinedChainExtension;
+
+/// Each chain extension has an abstract type([`CombinedChainExtension`]) that describes
+/// it and the actual instance that provides access to methods.
+/// This structure is an instance that is returned by the `self.env().extension()` call.
+///
+/// Because it is a combination of corresponding sub-instances, we need to initialize
+/// each sub-instance in the same way by calling [`ChainExtensionInstance::instantiate`].
+pub struct CombinedInstance {
+    /// The instance of the [`Psp22Extension`] chain extension.
+    ///
+    /// It provides you access to `PSP22` functionality.
+    pub psp22: <Psp22Extension as ChainExtensionInstance>::Instance,
+    /// The instance of the [`FetchRandom`] chain extension.
+    ///
+    /// It provides you access to randomness functionality.
+    pub rand: <FetchRandom as ChainExtensionInstance>::Instance,
+}
+
+impl ChainExtensionInstance for CombinedChainExtension {
+    type Instance = CombinedInstance;
+
+    fn instantiate() -> Self::Instance {
+        CombinedInstance {
+            psp22: <Psp22Extension as ChainExtensionInstance>::instantiate(),
+            rand: <FetchRandom as ChainExtensionInstance>::instantiate(),
+        }
+    }
+}
```

Comment:
The `instantiate()` method can be auto generated by macro since there is no custom logic to be implemented on the developer side. You already provide each extension type as `<Psp22Extension as ChainExtensionInstance>::Instance`. You can just automatically generate implementation through `Derive` macro.

```suggestion
#[derive(CombinedChainExtension)]
pub struct CombinedInstance {
    /// The instance of the [`Psp22Extension`] chain extension.
    ///
    /// It provides you access to `PSP22` functionality.
    pub psp22: <Psp22Extension as ChainExtensionInstance>::Instance,
    /// The instance of the [`FetchRandom`] chain extension.
    ///
    /// It provides you access to randomness functionality.
    pub rand: <FetchRandom as ChainExtensionInstance>::Instance,
}
```

Reply:
The `derive` macro can't modify the field's type. It leads to a strange API where the user still should use `<Psp22Extension as ChainExtensionInstance>::Instance`, but implementation will be generated automatically. So, I decided to go with the `macro_rule` approach. You can define the structure as usual structure, where each field is a chain extension, and it will produce the implementation for you.

```rust
ink::combine_extensions! {
    /// This extension combines the [`FetchRandom`] and [`Psp22Extension`] extensions.
    /// It is possible to combine any number of extensions in this way.
    ///
    /// This structure is an instance that is returned by the `self.env().extension()` call.
    pub struct CombinedChainExtension {
        /// The instance of the [`Psp22Extension`] chain extension.
        ///
        /// It provides you access to `PSP22` functionality.
        pub psp22: Psp22Extension,
        /// The instance of the [`FetchRandom`] chain extension.
        ///
        /// It provides you access to randomness functionality.
        pub rand: FetchRandom,
    }
}
```